nestkernel/stopwatch.h

```diff
@@ -0,0 +1,505 @@
+#include <sys/time.h>
+
+// C++ includes:
+#include "arraydatum.h"
+#include "dictdatum.h"
+#include "dictutils.h"
+#include <algorithm>
+#include <cassert>
+#include <chrono>
+#include <iostream>
+#include <vector>
+
+namespace nest
+{
+
+// TODO JV: Set this variable via cmake instead
+#ifdef TIMER_DETAILED
+constexpr bool use_detailed_timers = true;
+#else
+constexpr bool use_detailed_timers = false;
+#endif
+#ifdef THREADED_TIMERS
+constexpr bool use_threaded_timers = true;
+#else
+constexpr bool use_threaded_timers = false;
+#endif
+
+/********************************************************************************
+ * Stopwatch                                                                    *
+ *   Accumulates time between start and stop, and provides the elapsed time     *
+ *   with different time units. Either runs multi-threaded or only on master.   *
+ *                                                                              *
+ *   Usage example:                                                             *
+ *     Stopwatch< StopwatchVerbosity::Normal, StopwatchType::MasterOnly > x;    *
+ *     x.start();                                                               *
+ *     // ... do computations for 15.34 sec                                     *
+ *     x.stop(); // only pauses stopwatch                                       *
+ *     x.print("Time needed "); // > Time needed 15.34 sec.                     *
+ *     x.start(); // resumes stopwatch                                          *
+ *     // ... next computations for 11.22 sec                                   *
+ *     x.stop();                                                                *
+ *     x.print("Time needed "); // > Time needed 26,56 sec.                     *
+ *     x.reset(); // reset to default values                                    *
+ *     x.start(); // starts the stopwatch from 0                                *
+ *     // ... computation 5.7 sec                                               *
+ *     x.print("Time "); // > Time 5.7 sec.                                     *
+ *     // ^ intermediate timing without stopping the stopwatch                  *
+ *     // ... more computations 1.7643 min                                      *
+ *     x.stop();                                                                *
+ *     x.print("Time needed ", StopwatchBase< clock_type >::MINUTES, std::cerr);              *
+ *     // > Time needed 1,8593 min. (on cerr)                                   *
+ *     // other units and output streams possible                               *
+ ********************************************************************************/
+namespace timers
+{
+enum timeunit_t : size_t
+{
+  NANOSEC = 1,
+  MICROSEC = NANOSEC * 1000,
+  MILLISEC = MICROSEC * 1000,
+  SECONDS = MILLISEC * 1000,
+  MINUTES = SECONDS * 60,
+  HOURS = MINUTES * 60,
+  DAYS = HOURS * 24
+};
+
+template < clockid_t clock_type >
```

Comment:
Add  a doxygen comment for this class. Since it later is used to define variables called `..._timer_`, maybe call it `StopwatchTimer`?  That would also avoid confusion about the fact that this class is called `StopwatchBase` but then does *not* serve as base class for stopwatches.